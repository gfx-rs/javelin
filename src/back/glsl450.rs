use crate::{
    back::glsl_common::{write_type, Error, StatementBuilder},
    Binding, BuiltIn, FastHashMap, Module, TypeInner,
};
use std::{
    fmt::{self},
    io::Write,
};

pub fn write(module: &Module, out: &mut impl Write) -> Result<(), Error> {
    writeln!(out, "#version 450 core")?;

    let mut counter = 0;
    let mut names = FastHashMap::default();

    let mut namer = |name: Option<&String>| {
        if let Some(name) = name {
            names.insert(name.clone(), ());
            name.clone()
        } else {
            counter += 1;
            while names.get(&format!("_{}", counter)).is_some() {
                counter += 1;
            }
            format!("_{}", counter)
        }
    };

    let mut structs = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let name = namer(ty.name.as_ref());
                let mut idxs = Vec::new();

                for member in members {
                    idxs.push(namer(member.name.as_ref()))
                }

                structs.insert(handle, (name, idxs));
            }
            _ => continue,
        }
    }

    // Do a second pass to build the structs
    // TODO: glsl is order dependent so we need to build structs in order
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let (name, idxs) = structs.get(&handle).unwrap();

                writeln!(out, "struct {} {{", name)?;
                for (member, name) in members.iter().zip(idxs.iter()) {
                    writeln!(
                        out,
                        "   {} {}",
                        write_type(member.ty, &module.types, &structs)?,
                        name
                    )?;
                }
                writeln!(out, "}};")?;
            }
            _ => continue,
        }
    }

    let mut globals_lookup = FastHashMap::default();

    for (handle, global) in module.global_variables.iter() {
        if let Some(Binding::BuiltIn(built_in)) = global.binding {
            let semantic = match built_in {
                BuiltIn::Position => "gl_position",
                BuiltIn::GlobalInvocationId => "gl_GlobalInvocationID",
                BuiltIn::BaseInstance => "gl_BaseInstance",
                BuiltIn::BaseVertex => "gl_BaseVertex",
                BuiltIn::ClipDistance => "gl_ClipDistance",
                BuiltIn::InstanceIndex => "gl_InstanceIndex",
                BuiltIn::VertexIndex => "gl_VertexIndex",
                BuiltIn::PointSize => "gl_PointSize",
                BuiltIn::FragCoord => "gl_FragCoord",
                BuiltIn::FrontFacing => "gl_FrontFacing",
                BuiltIn::SampleIndex => "gl_SampleID",
                BuiltIn::FragDepth => "gl_FragDepth",
                BuiltIn::LocalInvocationId => "gl_LocalInvocationID",
                BuiltIn::LocalInvocationIndex => "gl_LocalInvocationIndex",
                BuiltIn::WorkGroupId => "gl_WorkGroupID",
            };

            globals_lookup.insert(handle, String::from(semantic));
            continue;
        }

        if let Some(ref binding) = global.binding {
            write!(out, "layout({}) ", binding.writer())?;
        }

        let name = namer(global.name.as_ref());

        writeln!(
            out,
            "{}{} {};",
            global.class.write_glsl(),
            write_type(global.ty, &module.types, &structs)?,
            name
        )?;

        globals_lookup.insert(handle, name);
    }

    let mut functions = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, func) in module.functions.iter() {
        functions.insert(handle, namer(func.name.as_ref()));
    }

    let mut types = module.types.clone();
    let mut typifier = crate::proc::Typifier::new();

    // TODO: glsl is order dependent so we need to build functions in order
    for (handle, func) in module.functions.iter() {
        let name = functions.get(&handle).unwrap();

        writeln!(
            out,
            "{} {}({}) {{",
            func.return_type
                .map_or(Ok(String::from("void")), |ty| write_type(
                    ty, &types, &structs
                ))?,
            name,
            func.parameter_types
                .iter()
                .map(|ty| write_type(*ty, &types, &structs))
                .collect::<Result<Vec<_>, _>>()?
                .join(","),
        )?;

        let mut builder = StatementBuilder {
            functions: &functions,
            globals: &globals_lookup,
            locals_lookup: &func
                .local_variables
                .iter()
                .map(|(handle, local)| (handle, namer(local.name.as_ref())))
                .collect(),
            structs: &structs,
            args: &func
                .parameter_types
                .iter()
                .enumerate()
                .map(|(pos, _)| (pos as u32, namer(None)))
                .collect(),
            expressions: &func.expressions,
            types: &mut types,
            locals: &func.local_variables,
            typifier: &mut typifier,
        };

        for sta in func.body.iter() {
            writeln!(out, "{}", sta.write_glsl(module, &mut builder)?)?;
        }

        writeln!(out, "}}")?;
    }

    Ok(())
}

struct BindingWriter<'a> {
    inner: &'a Binding,
}

impl Binding {
    fn writer<'a>(&'a self) -> BindingWriter<'a> {
        BindingWriter { inner: self }
    }
}

impl<'a> fmt::Display for BindingWriter<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.inner {
            Binding::BuiltIn(_) => write!(f, ""), // Ignore because they are variables with a predefined name
            Binding::Location(location) => write!(f, "location={}", location),
            Binding::Descriptor { set, binding } => write!(f, "set={},binding={}", set, binding),
        }
    }
}
