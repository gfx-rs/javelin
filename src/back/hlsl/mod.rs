mod keywords;
mod writer;

use std::fmt::Error as FmtError;
use thiserror::Error;

pub use writer::Writer;

#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
pub struct ShaderModel(u16);

impl Default for ShaderModel {
    fn default() -> Self {
        ShaderModel(50)
    }
}

#[derive(Error, Debug)]
pub enum Error {
    #[error(transparent)]
    IoError(#[from] FmtError),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedShaderModel(ShaderModel),
    #[error("A scalar with an unsupported width was requested: {0:?} {1:?}")]
    UnsupportedScalar(crate::ScalarKind, crate::Bytes),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedBuiltIn(crate::BuiltIn),
    #[error("{0}")]
    Unimplemented(String), // TODO: Error used only during development
}

pub fn write_string(
    module: &crate::Module,
    info: &crate::valid::ModuleInfo,
    shader_model: ShaderModel,
) -> Result<String, Error> {
    let mut w = Writer::new(String::new(), shader_model);
    w.write(module, info)?;
    let output = w.finish();
    Ok(output)
}
