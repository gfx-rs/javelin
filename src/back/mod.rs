//! Functions which export shader modules into binary and text formats.

#[cfg(feature = "glsl-out")]
pub mod glsl;
#[cfg(feature = "msl-out")]
pub mod msl;
#[cfg(feature = "spv-out")]
pub mod spv;

impl crate::Expression {
    /// Returns the ref count, upon reaching which this expression
    /// should be considered for baking.
    #[allow(dead_code)]
    fn bake_ref_count(&self) -> usize {
        match *self {
            // accesses are never cached, only loads are
            crate::Expression::Access { .. } | crate::Expression::AccessIndex { .. } => !0,
            // image operations look better when isolated
            crate::Expression::ImageSample { .. } | crate::Expression::ImageLoad { .. } => 1,
            // cache expressions that are referenced multiple times
            _ => 2,
        }
    }
}
