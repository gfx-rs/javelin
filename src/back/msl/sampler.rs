#[derive(Clone, Debug, PartialEq)]
pub enum Coord {
    Normalized,
    Pixel,
}

impl Default for Coord {
    fn default() -> Self {
        Self::Normalized
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Address {
    Repeat,
    MirroredRepeat,
    ClampToEdge,
    ClampToZero,
    ClampToBorder,
}

impl Default for Address {
    fn default() -> Self {
        Self::ClampToEdge
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum BorderColor {
    TransparentBlack,
    OpaqueBlack,
    OpaqueWhite,
}

impl Default for BorderColor {
    fn default() -> Self {
        Self::TransparentBlack
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Filter {
    Nearest,
    Linear,
}

impl Default for Filter {
    fn default() -> Self {
        Self::Nearest
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum CompareFunc {
    Never,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    Always,
}

impl Default for CompareFunc {
    fn default() -> Self {
        Self::Never
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct InlineSampler {
    pub coord: Coord,
    pub address: [Address; 3],
    pub border_color: BorderColor,
    pub mag_filter: Filter,
    pub min_filter: Filter,
    pub mip_filter: Option<Filter>,
    pub compare_func: CompareFunc,
}
