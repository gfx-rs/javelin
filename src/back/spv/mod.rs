mod layout;
pub mod parser;

use spirv::*;

struct Module {
    physical_layout: PhysicalLayout,
    logical_layout: LogicalLayout,
}

impl Module {
    fn new(header: &crate::Header) -> Self {
        Module {
            physical_layout: PhysicalLayout::new(header, 0x0u32),
            logical_layout: LogicalLayout::new(),
        }
    }
}

struct PhysicalLayout {
    magic_number: Word,
    version: Word,
    generator: Word,
    bound: Word,
    instruction_schema: Word,
}

impl PhysicalLayout {
    fn new(header: &crate::Header, instruction_schema: Word) -> Self {
        let version: Word = (0x0u32 << 24)
            | ((header.version.0 as u32) << 16)
            | ((header.version.1 as u32) << 8)
            | header.version.2 as u32;

        PhysicalLayout {
            magic_number: MAGIC_NUMBER,
            version,
            generator: header.generator,
            bound: 0,
            instruction_schema,
        }
    }

    fn in_words(&self) -> Vec<Word> {
        let mut words: Vec<Word> = vec![];

        words.push(self.magic_number);
        words.push(self.version);
        words.push(self.generator);

        match self.bound {
            0 => panic!("Bound cannot be 0"),
            _ => words.push(self.bound),
        }

        words.push(self.instruction_schema);
        words
    }
}

pub(crate) struct LogicalLayout {
    capabilities: Vec<Word>,
    extensions: Vec<Word>,
    ext_inst_imports: Vec<Word>,
    memory_model: Vec<Word>,
    entry_points: Vec<Word>,
    execution_modes: Vec<Word>,
    debugs: Vec<Word>,
    annotations: Vec<Word>,
    type_declarations: Vec<Word>,
    constants: Vec<Word>,
    global_variables: Vec<Word>,
    function_declarations: Vec<Word>,
    function_definitions: Vec<Word>,
}

impl LogicalLayout {
    fn new() -> Self {
        LogicalLayout {
            capabilities: vec![],
            extensions: vec![],
            ext_inst_imports: vec![],
            memory_model: vec![],
            entry_points: vec![],
            execution_modes: vec![],
            debugs: vec![],
            annotations: vec![],
            type_declarations: vec![],
            constants: vec![],
            global_variables: vec![],
            function_declarations: vec![],
            function_definitions: vec![],
        }
    }

    fn in_words(&self) -> Vec<u32> {
        let mut words: Vec<Word> = vec![];

        for capability in self.capabilities.iter() {
            words.push(*capability);
        }

        for extension in self.extensions.iter() {
            words.push(*extension);
        }

        for ext_inst_import in self.ext_inst_imports.iter() {
            words.push(*ext_inst_import);
        }

        for memory_model in self.memory_model.iter() {
            words.push(*memory_model);
        }

        for entry_point in self.entry_points.iter() {
            words.push(*entry_point);
        }

        for execution_mode in self.execution_modes.iter() {
            words.push(*execution_mode);
        }

        for debug in self.debugs.iter() {
            words.push(*debug);
        }

        for annotation in self.annotations.iter() {
            words.push(*annotation);
        }

        for type_declaration in self.type_declarations.iter() {
            words.push(*type_declaration);
        }

        for constants in self.constants.iter() {
            words.push(*constants);
        }

        for global_variables in self.global_variables.iter() {
            words.push(*global_variables);
        }

        for function_declaration in self.function_declarations.iter() {
            words.push(*function_declaration);
        }

        for function_definition in self.function_definitions.iter() {
            words.push(*function_definition);
        }

        words
    }
}

struct Instruction {
    op: Op,
    wc: u32,
    type_id: Option<Word>,
    result_id: Option<Word>,
    operands: Vec<Word>,
}

impl Instruction {
    fn new(op: Op) -> Self {
        Instruction {
            op,
            wc: 1, // Always start at 1 for the first word (OP + WC),
            type_id: None,
            result_id: None,
            operands: vec![],
        }
    }

    fn set_type(&mut self, id: Word) {
        assert!(self.type_id.is_none(), "Type can only be set once");
        self.type_id = Some(id);
        self.wc += 1;
    }

    fn set_result(&mut self, id: Word) {
        assert!(self.result_id.is_none(), "Result can only be set once");
        self.result_id = Some(id);
        self.wc += 1;
    }

    fn add_operand(&mut self, operand: Word) {
        self.operands.push(operand);
        self.wc += 1;
    }

    fn add_operands(&mut self, operands: Vec<Word>) {
        for operand in operands {
            self.add_operand(operand);
        }
    }

    fn to_words(&self) -> Vec<Word> {
        let mut words = Vec::with_capacity(self.wc as usize);

        let wc_op = (self.wc << 16 | self.op as u32) as u32;

        words.push(wc_op);

        if self.type_id.is_some() {
            words.push(self.type_id.unwrap());
        }

        if self.result_id.is_some() {
            words.push(self.result_id.unwrap());
        }

        for operand in self.operands.iter() {
            words.push(*operand);
        }

        words
    }
}
