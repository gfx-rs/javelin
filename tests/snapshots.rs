use serde::Deserialize;

#[derive(Hash, PartialEq, Eq, Deserialize)]
enum Stage {
    Vertex,
    Fragment,
    Compute,
}

#[derive(Hash, PartialEq, Eq, Deserialize)]
struct BindSource {
    stage: Stage,
    group: u32,
    binding: u32,
}

#[derive(Deserialize)]
struct BindTarget {
    #[serde(default)]
    buffer: Option<u8>,
    #[serde(default)]
    texture: Option<u8>,
    #[serde(default)]
    sampler: Option<u8>,
    #[serde(default)]
    mutable: bool,
}

#[derive(Default, Deserialize)]
struct Parameters {
    #[serde(default)]
    spv_flow_dump_prefix: String,
    #[cfg_attr(not(feature = "spv-out"), allow(dead_code))]
    spv_capabilities: naga::FastHashSet<spirv::Capability>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    mtl_bindings: naga::FastHashMap<BindSource, BindTarget>,
}

#[cfg(feature = "msl-out")]
fn check_output_msl(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::msl;
    let mut binding_map = msl::BindingMap::default();
    for (key, value) in params.mtl_bindings.iter() {
        binding_map.insert(
            msl::BindSource {
                stage: match key.stage {
                    Stage::Vertex => naga::ShaderStage::Vertex,
                    Stage::Fragment => naga::ShaderStage::Fragment,
                    Stage::Compute => naga::ShaderStage::Compute,
                },
                group: key.group,
                binding: key.binding,
            },
            msl::BindTarget {
                buffer: value.buffer,
                texture: value.texture,
                sampler: value.sampler,
                mutable: value.mutable,
            },
        );
    }
    let options = msl::Options {
        lang_version: (1, 0),
        spirv_cross_compatibility: false,
        binding_map,
    };
    let (msl, _) = msl::write_string(&module, &options).unwrap();
    insta::assert_snapshot!(format!("{}{}", name, ".msl"), msl);
}

#[cfg(feature = "spv-out")]
fn check_output_spv(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::spv;
    use rspirv::binary::Disassemble;
    let spv = spv::write_vec(
        &module,
        spv::WriterFlags::NONE,
        params.spv_capabilities.clone(),
    )
    .unwrap();
    let dis = rspirv::dr::load_words(spv)
        .expect("Produced invalid SPIR-V")
        .disassemble();
    insta::assert_snapshot!(format!("{}{}", name, ".spvasm"), dis);
}

#[cfg(feature = "wgsl-in")]
fn convert_wgsl(name: &str) {
    let params =
        match std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".param.ron")) {
            Ok(string) => ron::de::from_str(&string).expect("Couldn't find param file"),
            Err(_) => Parameters::default(),
        };

    let module = naga::front::wgsl::parse_str(
        &std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".wgsl"))
            .expect("Couldn't find wgsl file"),
    )
    .unwrap();
    naga::proc::Validator::new().validate(&module).unwrap();

    #[cfg(feature = "msl-out")]
    check_output_msl(&module, name, &params);
    #[cfg(feature = "spv-out")]
    check_output_spv(&module, name, &params);
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_quad() {
    convert_wgsl("quad");
}
