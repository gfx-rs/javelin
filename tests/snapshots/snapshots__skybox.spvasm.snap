---
source: tests/snapshots.rs
expression: dis
---
; SPIR-V
; Version: 1.0
; Generator: rspirv
; Bound: 185
OpCapability Shader
%1 = OpExtInstImport "GLSL.std.450"
OpMemoryModel Logical GLSL450
OpEntryPoint Vertex %11 "vs_main" %157 %44 %15
OpEntryPoint Fragment %168 "fs_main" %180 %170
OpExecutionMode %168 OriginUpperLeft
OpDecorate %15 BuiltIn VertexIndex
OpDecorate %26 Block
OpMemberDecorate %26 0 Offset 0
OpMemberDecorate %26 0 MatrixStride 16
OpDecorate %27 ColMajor
OpMemberDecorate %26 1 Offset 64
OpMemberDecorate %26 1 MatrixStride 16
OpDecorate %25 DescriptorSet 0
OpDecorate %25 Binding 0
OpDecorate %44 Location 0
OpDecorate %48 ColMajor
OpDecorate %157 BuiltIn Position
OpDecorate %170 Location 0
OpDecorate %171 DescriptorSet 0
OpDecorate %171 Binding 1
OpDecorate %176 DescriptorSet 0
OpDecorate %176 Binding 2
OpDecorate %180 Location 0
%3 = OpTypeInt 32 1
%4 = OpTypePointer Function %3
%8 = OpTypeFloat 32
%7 = OpTypeVector %8 4
%9 = OpTypePointer Function %7
%10 = OpTypeVoid
%12 = OpTypeFunction %10
%16 = OpTypeInt 32 0
%17 = OpTypePointer Input %16
%15 = OpVariable  %17  Input
%19 = OpConstant  %3  2
%22 = OpConstant  %3  1
%27 = OpTypeMatrix %7 4
%26 = OpTypeStruct %27 %27
%28 = OpTypePointer Uniform %26
%25 = OpVariable  %28  Uniform
%29 = OpTypePointer Uniform %27
%30 = OpConstant  %3  0
%36 = OpConstant  %8  4.0
%37 = OpConstant  %8  1.0
%42 = OpConstant  %8  0.0
%45 = OpTypeVector %8 3
%46 = OpTypePointer Output %45
%44 = OpVariable  %46  Output
%48 = OpTypeMatrix %45 3
%52 = OpTypePointer Uniform %27
%53 = OpConstant  %3  1
%54 = OpTypePointer Uniform %7
%55 = OpConstant  %3  0
%56 = OpTypePointer Uniform %8
%57 = OpConstant  %3  0
%62 = OpTypePointer Uniform %27
%63 = OpConstant  %3  1
%64 = OpTypePointer Uniform %7
%65 = OpConstant  %3  0
%66 = OpTypePointer Uniform %8
%67 = OpConstant  %3  1
%72 = OpTypePointer Uniform %27
%73 = OpConstant  %3  1
%74 = OpTypePointer Uniform %7
%75 = OpConstant  %3  0
%76 = OpTypePointer Uniform %8
%77 = OpConstant  %3  2
%83 = OpTypePointer Uniform %27
%84 = OpConstant  %3  1
%85 = OpTypePointer Uniform %7
%86 = OpConstant  %3  1
%87 = OpTypePointer Uniform %8
%88 = OpConstant  %3  0
%93 = OpTypePointer Uniform %27
%94 = OpConstant  %3  1
%95 = OpTypePointer Uniform %7
%96 = OpConstant  %3  1
%97 = OpTypePointer Uniform %8
%98 = OpConstant  %3  1
%103 = OpTypePointer Uniform %27
%104 = OpConstant  %3  1
%105 = OpTypePointer Uniform %7
%106 = OpConstant  %3  1
%107 = OpTypePointer Uniform %8
%108 = OpConstant  %3  2
%114 = OpTypePointer Uniform %27
%115 = OpConstant  %3  1
%116 = OpTypePointer Uniform %7
%117 = OpConstant  %3  2
%118 = OpTypePointer Uniform %8
%119 = OpConstant  %3  0
%124 = OpTypePointer Uniform %27
%125 = OpConstant  %3  1
%126 = OpTypePointer Uniform %7
%127 = OpConstant  %3  2
%128 = OpTypePointer Uniform %8
%129 = OpConstant  %3  1
%134 = OpTypePointer Uniform %27
%135 = OpConstant  %3  1
%136 = OpTypePointer Uniform %7
%137 = OpConstant  %3  2
%138 = OpTypePointer Uniform %8
%139 = OpConstant  %3  2
%145 = OpTypePointer Function %8
%146 = OpConstant  %3  0
%149 = OpTypePointer Function %8
%150 = OpConstant  %3  1
%153 = OpTypePointer Function %8
%154 = OpConstant  %3  2
%158 = OpTypePointer Output %7
%157 = OpVariable  %158  Output
%170 = OpVariable  %158  Output
%172 = OpTypeImage %8 Cube 0 0 0 1 Unknown
%173 = OpTypePointer UniformConstant %172
%171 = OpVariable  %173  UniformConstant
%175 = OpTypeSampledImage %172
%177 = OpTypeSampler
%178 = OpTypePointer UniformConstant %177
%176 = OpVariable  %178  UniformConstant
%181 = OpTypePointer Input %45
%180 = OpVariable  %181  Input
%11 = OpFunction  %10  None %12
%13 = OpLabel
%2 = OpVariable  %4  Function
%5 = OpVariable  %4  Function
%6 = OpVariable  %9  Function
%18 = OpLoad  %16  %15
%14 = OpSDiv  %3  %18 %19
OpStore %2 %14
%21 = OpLoad  %16  %15
%20 = OpBitwiseAnd  %3  %21 %22
OpStore %5 %20
%24 = OpAccessChain  %29  %25 %30
%31 = OpLoad  %27  %24
%34 = OpLoad  %3  %2
%35 = OpConvertSToF  %8  %34
%33 = OpFMul  %8  %35 %36
%32 = OpFSub  %8  %33 %37
%40 = OpLoad  %3  %5
%41 = OpConvertSToF  %8  %40
%39 = OpFMul  %8  %41 %36
%38 = OpFSub  %8  %39 %37
%43 = OpCompositeConstruct  %7  %32 %38 %42 %37
%23 = OpMatrixTimesVector  %7  %31 %43
OpStore %6 %23
%51 = OpAccessChain  %52  %25 %53
%50 = OpAccessChain  %54  %51 %55
%49 = OpAccessChain  %56  %50 %57
%58 = OpLoad  %8  %49
%61 = OpAccessChain  %62  %25 %63
%60 = OpAccessChain  %64  %61 %65
%59 = OpAccessChain  %66  %60 %67
%68 = OpLoad  %8  %59
%71 = OpAccessChain  %72  %25 %73
%70 = OpAccessChain  %74  %71 %75
%69 = OpAccessChain  %76  %70 %77
%78 = OpLoad  %8  %69
%79 = OpCompositeConstruct  %45  %58 %68 %78
%82 = OpAccessChain  %83  %25 %84
%81 = OpAccessChain  %85  %82 %86
%80 = OpAccessChain  %87  %81 %88
%89 = OpLoad  %8  %80
%92 = OpAccessChain  %93  %25 %94
%91 = OpAccessChain  %95  %92 %96
%90 = OpAccessChain  %97  %91 %98
%99 = OpLoad  %8  %90
%102 = OpAccessChain  %103  %25 %104
%101 = OpAccessChain  %105  %102 %106
%100 = OpAccessChain  %107  %101 %108
%109 = OpLoad  %8  %100
%110 = OpCompositeConstruct  %45  %89 %99 %109
%113 = OpAccessChain  %114  %25 %115
%112 = OpAccessChain  %116  %113 %117
%111 = OpAccessChain  %118  %112 %119
%120 = OpLoad  %8  %111
%123 = OpAccessChain  %124  %25 %125
%122 = OpAccessChain  %126  %123 %127
%121 = OpAccessChain  %128  %122 %129
%130 = OpLoad  %8  %121
%133 = OpAccessChain  %134  %25 %135
%132 = OpAccessChain  %136  %133 %137
%131 = OpAccessChain  %138  %132 %139
%140 = OpLoad  %8  %131
%141 = OpCompositeConstruct  %45  %120 %130 %140
%142 = OpCompositeConstruct  %48  %79 %110 %141
%143 = OpTranspose  %48  %142
%144 = OpAccessChain  %145  %6 %146
%147 = OpLoad  %8  %144
%148 = OpAccessChain  %149  %6 %150
%151 = OpLoad  %8  %148
%152 = OpAccessChain  %153  %6 %154
%155 = OpLoad  %8  %152
%156 = OpCompositeConstruct  %45  %147 %151 %155
%47 = OpMatrixTimesVector  %45  %143 %156
OpStore %44 %47
%161 = OpLoad  %3  %2
%162 = OpConvertSToF  %8  %161
%160 = OpFMul  %8  %162 %36
%159 = OpFSub  %8  %160 %37
%165 = OpLoad  %3  %5
%166 = OpConvertSToF  %8  %165
%164 = OpFMul  %8  %166 %36
%163 = OpFSub  %8  %164 %37
%167 = OpCompositeConstruct  %7  %159 %163 %42 %37
OpStore %157 %167
OpReturn
OpFunctionEnd
%168 = OpFunction  %10  None %12
%169 = OpLabel
%174 = OpLoad  %172  %171
%179 = OpLoad  %177  %176
%182 = OpLoad  %45  %180
%183 = OpSampledImage  %175  %174 %179
%184 = OpImageSampleImplicitLod  %7  %183 %182
OpStore %170 %184
OpReturn
OpFunctionEnd
